src/dynamic_honey_badger/votes.rs

```diff
@@ -0,0 +1,302 @@
+use messaging::NetworkInfo;
+use std::collections::{btree_map, BTreeMap, HashMap};
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::rc::Rc;
+
+use bincode;
+use serde::{Deserialize, Serialize};
+
+use super::{Change, Result};
+use crypto::Signature;
+use fault_log::{Fault, FaultKind, FaultLog};
+
+/// A buffer and counter collecting pending and committed votes for validator set changes.
+///
+/// This is reset whenever the set of validators changes or a change reaches a majority. We call
+/// the epochs since the last reset the current _era_.
+pub struct VoteCounter<NodeUid> {
+    /// Shared network data.
+    netinfo: Rc<NetworkInfo<NodeUid>>,
+    /// The epoch when voting was reset.
+    era: u64,
+    /// Pending node transactions that we will propose in the next epoch.
+    pending: BTreeMap<NodeUid, SignedVote<NodeUid>>,
+    /// Collected votes for adding or removing nodes. Each node has one vote, and casting another
+    /// vote revokes the previous one.
+    committed: BTreeMap<NodeUid, Vote<NodeUid>>,
+}
+
+impl<NodeUid> VoteCounter<NodeUid>
+where
+    NodeUid: Eq + Hash + Ord + Clone + Debug + Serialize + for<'r> Deserialize<'r>,
+{
+    /// Creates a new `VoteCounter` object with empty buffer and counter.
+    pub fn new(netinfo: Rc<NetworkInfo<NodeUid>>, era: u64) -> Self {
+        VoteCounter {
+            era,
+            netinfo,
+            pending: BTreeMap::new(),
+            committed: BTreeMap::new(),
+        }
+    }
+
+    /// Creates a signed vote for the given change, and inserts it into the pending votes buffer.
+    pub fn sign_vote_for(&mut self, change: Change<NodeUid>) -> Result<&SignedVote<NodeUid>> {
+        let voter = self.netinfo.our_uid().clone();
+        let vote = Vote {
+            change,
+            era: self.era,
+            num: self.pending.get(&voter).map_or(0, |sv| sv.vote.num + 1),
+        };
+        let ser_vote = bincode::serialize(&vote)?;
+        let signed_vote = SignedVote {
+            vote,
+            voter: voter.clone(),
+            sig: self.netinfo.secret_key().sign(ser_vote),
+        };
+        self.pending.insert(voter.clone(), signed_vote);
+        Ok(self.pending.get(&voter).expect("entry was just inserted"))
+    }
+
+    /// Inserts a pending vote into the buffer, if it has a higher number than the existing one.
+    pub fn add_pending_vote(
+        &mut self,
+        sender_id: &NodeUid,
+        signed_vote: SignedVote<NodeUid>,
+    ) -> Result<FaultLog<NodeUid>> {
+        if signed_vote.vote.era != self.era {
+            return Ok(FaultLog::new()); // The vote is obsolete.
+        }
+        if !self.validate(&signed_vote)? {
+            return Ok(Fault::new(sender_id.clone(), FaultKind::InvalidVoteSignature).into());
```

Comment:
You could remove the call to `.into()` by using `Ok(FaultLog::init(sender_id.clone(), FaultKind::InvalidVoteSignature))`.

Reply:
Same thing goes for everywhere else this is used. Both options do exactly the same thing though.